synapse/util/caches/deferred_cache.py

```diff
@@ -214,16 +214,25 @@ def set(
 
         callbacks = [callback] if callback else []
         self.check_thread()
-        observable = ObservableDeferred(value, consumeErrors=True)
-        observer = observable.observe()
-        entry = CacheEntry(deferred=observable, callbacks=callbacks)
 
         existing_entry = self._pending_deferred_cache.pop(key, None)
         if existing_entry:
             existing_entry.invalidate()
 
         # XXX: why don't we invalidate the entry in `self.cache` yet?
 
+        # we can save a whole load of effort if the deferred is ready.
+        if value.called:
+            self.cache.set(key, value.result, callbacks)
+            return value
```

Comment:
I'm not really sure it matters, but switching from returning  `ObservableDeferred.observe()` to `value` means that the result of this `Deferred` could now be eaten in the callback chain where previously it wouldn't be. Perhaps this should be `return defer.succeed(value)`?

Reply:
I don't think it matters if it gets eaten further down the callback chain? we've got the result and done everything we need with it?

>  Perhaps this should be return defer.succeed(value)?

`value` is itself a Deferred, so possibly `defer.succeed(value.result)`, but I'm not sure it's necessary.

(Tests might be fixed now)

Reply:
I'm trying to come up with a situation in which it would matter and I'm struggling to find any. It might matter if the caller is setting the cache and doing other things with the result, but I'm not sure that would make much sense. 🤷